### ISSUE TITLE
AR Core / Width prediction wrong / Put warning to users.

### ISSUE THREAD

**Issue by flohoff** [streetcomplete/StreetComplete#4655]
Hi,
i found multiple new introduced routing problems which were caused by people using the Width estimation. These popped up in my Routing Quality Assurance where i calculate 200K+ routes every couple of hours for a given set of destinations.

The users introduced widths of 1.6-2.5m on streets (All oneway, single lane, 4.5m+ width). This caused routing issues as roads with <= 1.9 are considered to be impassable by cars, and roads <2.2m are hit by a significant penalty in OSRM.

I would consider a big fat warning popup when people put in something <2.6m on roads as these will be immediatly impassable by HGV and Busses (width 2.55m), and definitly refuse to put something <2m on highways which is broken in 99% of the cases. 

In this case the roads contain Bus routes. Busses are definitly >=2.50.

Flo

**Comment by flohoff** [comment]
Basically all measurements the user made were wrong. At least those i found. They were all approx halve the correct values i'd guess. As the user is a former colleague of mine i contacted him asking which App he used. I talked a while and he didnt even notice the completely bogus widths. 

I removed/reverted the changes.

Changeset/Ways i remember:

https://www.openstreetmap.org/way/777159350
https://www.openstreetmap.org/way/34128448
https://www.openstreetmap.org/changeset/129437713

Basically those widths made the ways unusable. 1.6m is completely out of scope of public streets, widths <2.6m are not for Busses, HGV and stuff.  

So all values for public roads <3m i would have serious doubts.

Flo

**Comment by flohoff** [comment]
Here are mapillary Images of the locations - Both are oneway, the second one has trolley/bus tours running - From good aerial i would guess both road segments are 4m+ (Josm line length measurements on 20cm Aerial imagery)

https://www.mapillary.com/app/?pKey=222366242600068
https://www.mapillary.com/app/?pKey=813682756247703

Flo

**Comment by paulinsche** [comment]
> I talked a while and he didnt even notice the completely bogus widths.

Next time, I'll take a look at whether and how I've managed to operate it incorrectly.

**Comment by westnordost** [comment]
The further the arrow is away from yourself (i.e. the further you rotate the smartphone towards the horizon rather than towards the ground), the higher the imprecision, because trigonometry.

However, that shouldn't result in measurements that are so completely off and if I remember correctly, the app will actually not allow you to point the smartphone so far away, displaying a message like "keep closer to the arrow".
In general, the most precise measurements are made if you start the measurement by pointing the smartphone down, then crossing the road following the arrow and then again complete the measurement by pointing the smartphone down.

Anyway, the proposed solution with the warning dialog for unrealistic road widths make sense. It should also be added to cycleway and maxwidth (chokers etc) widths.

**Comment by paulinsche** [comment]
> The further the arrow is away from yourself (i.e. the further you rotate the smartphone towards the horizon rather than towards the ground), the higher the imprecision, because trigonometry.

But that is exactly how I proceeded.

If I was presented with instructions, I skipped them.

**Comment by westnordost** [comment]
How did you proceed?

**Comment by paulinsche** [comment]
I stood at a kerb, held the mobile phone down and fixed it with a red dot. Then I let the mobile phone point in the direction of the opposite kerb and fixed it. I entered the value with full confidence in the technology. In fact, I did not pay attention to the values.

**Comment by westnordost** [comment]
I see, so an additional information dialog should be shown here too.

In my tests, doing that was basically never possible because ARCore would not detect the horizontal plane that far away from the camera picture but the arrow can only move on that horizontal plane. The arrow can only ever be on the area detected as a the horizontal plane (AKA "the ground"). If you point your smartphone e.g. into the sky, or for that matter outside of that area, a message like "keep closer to the arrow" was displayed.

I am interested in the % of error introduced depending on the method used for measuring. Could you test and compare the two methods on the ground and see by how much they differ? Also, I wonder, in any of these two methods, is the arrow floating in the air and not actually on the ground? (You can check this by just lowering your smartphone)

(Also if you have a laser distance meter or tape measure, I'd also be interested in the difference between the measurement with the smartphone and the real width. On my tests, the error was about 1-3% if I remember correctly, i.e. about 10-20cm on a 8m wide road)

**Comment by paulinsche** [comment]
> I am interested in the % of error introduced depending on the method used for measuring. Could you test and compare the two methods on the ground and see by how much they differ?

I'll try my very best.

**Comment by flohoff** [comment]
Still - This type of accidential use of the width measurement really breaks routing.

We need to protect the OSM Data against accidential breakage. Setting stuff <3m on and highway=residential or better is in 99.9% an error and the app should prevent this from happing in the first place.

In this case i could detect it pretty fast (<24h) as routing changed and i am monitoring in the East of Northrhine-Westfalia. Still this app now sprinkles breakage globally which will go undetected for a very long time.

Flo

**Comment by mnalis** [comment]
That is indeed worrisome that incorrect `width` does real-world routing damage (I wasn't aware it was _actually_ being used in practice). **Especially** since I'm guessing many other users may not be paying attention to actual results either, but just trust in the [technology just working](https://github.com/streetcomplete/StreetComplete/issues/4655#issuecomment-1331207478).

- Because it is causing real damage (which is even in relatively small amounts probably more problematic then potential usefulness gains), combined with (**[ongoing!](https://github.com/streetcomplete/StreetComplete/issues/4289#issuecomment-1266785943)**) licensing issues in https://github.com/streetcomplete/StreetComplete/issues/4289, I'm wondering if that AR experiment should be proclaimed failed and **_AR functionality removed_** / disabled?

- Alternatively (or as a quick workaround while we debate), perhaps we should at least start tagging them with [est_width=*](https://wiki.openstreetmap.org/wiki/Key:est_width) **_instead of_** [width=*](https://wiki.openstreetmap.org/wiki/Key:width) (especially given that **both** AR technology is lacking, and **that** non-AR measurements would more likely often be a **guess** then people carrying a laser distance-meter with them when mapping) ?

**Comment by westnordost** [comment]
Such a step needs to be backed up by actual numbers on the mean precision of this data and the prevalence of (extreme) outliers due to whatever caused the massive imprecision to be around 50% instead of the 1-3% in the changeset found.

**Comment by mnalis** [comment]
It might be quite a research project, though :face_with_head_bandage: - there are [17k+ samples](https://taginfo.openstreetmap.org/tags/source%3Awidth=ARCore#overview), but even if much reduced set size is taken at random, probably at least a few hundreds should be checked before we get toward any statistical meaningfullness.

Quick [overpass](http://overpass-turbo.eu/s/1onX) shows (not counting my own, which I know are correct-ish as I did pay attention to the values and used AR in a most recommended way) just one other use in Croatia, which happens to be wrong:
https://www.openstreetmap.org/way/712502453 says `5.9m`, when measuring via 1:5000 aerial gives about `3.2m` or `8.0m` - because it's actually **two** different-sized roads in L-shape, but the user apparently ignored that (_in addition_ to AR measuring incorrectly!) . :cry: 

I know, [anecdotal evidence](https://www.wordnik.com/words/anecdotal%20evidence), but that is exactly why sample size should be much larger (and include different users in different locations etc.) 

Anyway, if anyone has time to do more research (and either can verify in person, or has good-resolution aerials to measure in JOSM, or even better has access to city planners data specifying exact widths), here is [overpass query](http://overpass-turbo.eu/s/1onY)  which shows ARcore-measured roads which are not between `2.5m` - `5.0m` (just an example, easily changeable in the query).

Or perhaps we should _disable the quest by default_ as a precaution until someone decides to do such research project on bigger sample?  (That at least would insure that persons mapping the widths have expanded effort into enabling the quest, so are more likely to actually verify if the result looks reasonable instead of accepting whatever app returns as result)

**Comment by westnordost** [comment]
Jesus, Matja, don't panic. How about we first try to find out why these measurements were that far off and see if we can prevent that through UX or technical measures if there is anything systematic about it, like we do with any other case when people report that wrong data has been recorded with this app?

**Comment by flohoff** [comment]
Here is the assumed car profile width in the OSRM Profile for impassable roads:

https://github.com/Project-OSRM/osrm-backend/blob/2044398dfdab46dd13e6c0a17df76ffb9a6279d2/profiles/car.lua#L44

This is the width penalty if the road is <3m and is bidirectional the assumed penalty is 50% or halve the assumed speed from maxspeed:

https://github.com/Project-OSRM/osrm-backend/blob/2044398dfdab46dd13e6c0a17df76ffb9a6279d2/profiles/lib/way_handlers.lua#L400

My guess is that most navigational engines have profile informations like that. The max width of a Bus/HGV in Europe is 2.55m. So when tagging something <2.60 most likely all navigational profiles for Bus/HGV will ignore this road.

The issue i have here that width is most likely one of the most dangerous tags to set. We dont have global routing Quality Assurance so by tagging just 1 per thousands smaller than 2.55 will seriously deteriorate the routing capabilities in OSM over the long term.

These bugs can basically not be found manually just by looking at the data and only be reverified by going on site.


So my suggestion is still to refuse tagging anything smaller than 2m on a highway tagged road for cars ( path, cycleway, footway would be okay ) with the ARCore width, and make a big fat popup that this is a highly suspicious value if <3m.


Flo

**Comment by westnordost** [comment]
Yes,  I agree, a warning dialog for unrealistic values should be added.  This ticket has been tagged with "enhancement" which means that the suggestion has been acknowledged to make sense. 

The discussion with mnalis went beyond that about whether the mismeasurement is a systematic error (after all if mismeasurement can happen with narrow roads, it can happen even more with broad roads). If it is,  additionally it is even more important to fix that.  After it is found out how this is caused.

**Comment by westnordost** [comment]
@paulinsche were you able to reproduce the issue?

**Comment by westnordost** [comment]
Leaving open, see https://github.com/streetcomplete/StreetComplete/issues/4655#issuecomment-1331988493 and following comments.

@flohoff here is the code that decides whether a specified width is dubious, maybe this is useful to your QA tool, too. In particular, notice the `estimateUsableRoadwayWidth` function defined in the same file:

https://github.com/streetcomplete/StreetComplete/blob/a36e2fb5942118a42027b35f06da127d1943ae3a/app/src/main/java/de/westnordost/streetcomplete/osm/RoadWidth.kt#L73-L104

**Comment by flohoff** [comment]
I do QA by analysing routing differences over time. So if a route changes i compare old and new route geometry. I rely on OSRM calculating routes. OSRM considers roads <1.9m to be impassable by cars. So your <1.75 still has the chance of making track/service roads impassable by the OSRM default car profile.

**Comment by westnordost** [comment]
I can imagine, but it is not uncommon for `service` roads, especially driveways, to be somewhat narrow. But anyway, let's increase that to 1.9m then. After all, pretty much all SUVs are somewhat broader than what is usually assumed to be the default car width.

**Comment by westnordost** [comment]
So the ARcore stuff is being outsourced into an own app. I took this as an opportunity to add the following info dialog that explains how the best precision can be achieved:

![Screenshot_20230105_004242_StreetMeasure](https://user-images.githubusercontent.com/4661658/210670814-8dce72ae-5bf4-4a06-b3e4-8ae5f468ae26.jpg)

Maybe I'll also outright disallow taking measurements if the tilt is above a certain angle

**Comment by paulinsche** [comment]
> 

Sorry for late response. I got following results on a 5.5m street (mit Zollstock): 

Min: 3.5m, max: 4.5m

Still do not understanding, what I'm doing wrong.

**Comment by westnordost** [comment]
Does the above text in the screenshot  help you? 

On January 5, 2023 1:31:39 AM GMT+01:00, paulinsche ***@***.***> wrote:
>> 
>
>Sorry for late response. I got following results on a 5.5m street (mit Zollstock): 
>
>Min: 3.5m, max: 4.5m
>
>Still do not understanding, what I'm doing wrong.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/issues/4655#issuecomment-1371582820
>You are receiving this because you were assigned.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by westnordost** [comment]
I also explained it in more detail here with illustration: https://github.com/MarcusWolschon/osmeditor4android/issues/2010#issuecomment-1368210588

**Comment by paulinsche** [comment]
> Does the above text in the screenshot help you?

No. But I'll read the Text with the Illustration.

**Comment by SomeoneElseOSM** [comment]
Separately from the issue about ARCore width prediction problems, I think that it's important to comment on some of the assumptions made in the first post here.

> I would consider a big fat warning popup when people put in something <2.6m on roads as these will be immediatly impassable by HGV and Busses (width 2.55m),

I can think of plenty of narrow roads near me (Yorkshire, England) which are less than 2.6m wide but regularly get HGV traffic (many farm tracks).  The reason, of course, as that the vehicle's track is less than or about the same as the road width, but the hedges are set back from the road allowing HGV traffic (often regular, in the case of milk tankers).  Roads that are part of the public road system that are ~2m wide are rarer, but do exist.  Most do not explicitly exclude HGVs unless there's an actual width restriction due to walls or buildings that limits the width of vehicles that can pass.

> In this case the roads contain Bus routes. Busses are definitly >=2.50.

Something that might be tagged as a bus route in some places outside Europe might not be serviced by something you would immediately classified as a bus. Hong Kong's "minibus" classification is one example that comes immediately to mind; scheduled minibus taxis in Africa may also fit.

Even in Europe, taking the Yorkshire example again, the bit of the bus route at https://www.openstreetmap.org/relation/2993976#map=15/54.0939/-1.1093 have road widths around 3m, with some bits less, yet are served by regular UK single-decker buses.  Again, the bus fits because the track <= road width and the width between hedges is > bus width (just, in this case).

**Comment by flohoff** [comment]
> I can think of plenty of narrow roads near me (Yorkshire, England) which
> are less than 2.6m wide but regularly get HGV traffic (many farm tracks). The
> reason, of course, as that the vehicle's track is less than or about the same as the road
> width, but the hedges are set back from the road allowing HGV traffic (often regular,
> in the case of milk tankers). Roads that are part of the public road system that are ~2m 
> wide are rarer, but do exist. Most do not explicitly exclude HGVs unless there's an
>  actual width restriction due to walls or buildings that limits the width of vehicles that 
> can pass.

This is why i proposed a warning not a refusal. With farm traffic (As i happen to live in VERY rural Germany) is most like even a lot wider that 2.55 which is the legal limit in Europe for street legal cars/hgv with the exemption of farm vehicles. So is is not rare that farm vehicles exceed 3m. So when you tag a road with 1.6m i doubt that even with hedges you want to traverse that road with a farm vehicle or even a car (My car is 2,20m with mirrors so i would not try to put it on that road)

Tagging stuff <=1.9m is breaking for Car routing, if tagging <=2.55m its breaking HGV routing. And the mapper should be made aware of this.

And thats the definition of the width tag on roads. Its the physical limit which must not be exceeded. If you decide that the track is just 1.6m on the ground, and 3m at 50cm hight, you may decide to tag 3m. But using coreai for width estimation is unusable in that case anyway. And we are only talking about a warning popup if coreai width estimation returns very narrow roads. 

Flo

**Comment by SomeoneElseOSM** [comment]
> And we are only talking about a warning popup if coreai width estimation returns very narrow roads.

Whilst this is what this issue text originally suggested, that's not how it has been implemented in StreetComplete - I've seen a warning pop up when not using ARCore to measure the width or a narrow road after measuring or estimating it (there may of course be valid reasons due to other reports to add that warning more generally, of course).

**Comment by westnordost** [comment]
Well, I think it is still perfectly reasonable to ask for confirmation for unusual/unexpectedly narrow widths. After all, you can make mistakes measuring with any other measure, too.

**Comment by SomeoneElseOSM** [comment]
The way that I read https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/osm/RoadWidth.kt#L73 is that any residential road < 5.2m wide will generate the prompt (I've just had the prompt on a 4m wide one).  This seems ... a bit excessive.

**Comment by westnordost** [comment]
A width of below 5.2m on a non-oneway road means that two oncoming cars cannot pass each other.

**Comment by mnalis** [comment]
- ref: https://github.com/streetcomplete/StreetComplete/discussions/4742
